src/main/app/index.js

```diff
@@ -231,25 +273,134 @@ class App {
+    for (const { isDir, path } of pathsToOpen) {
+      if (isDir) {
+        directorySet.add(path)
+      } else {
+        fileSet.add(path)
+      }
+    }
+
+    // Filter out directories that are already opened.
+    for (const window of _windowManager.windows.values()) {
+      if (window.type === WindowType.EDITOR) {
+        const { openedRootDirectory } = window
+        if (directorySet.has(openedRootDirectory)) {
+          window.bringToFront()
+          directorySet.delete(openedRootDirectory)
+        }
+      }
+    }
+
+    const directoriesToOpen = Array.from(directorySet).map(dir => ({ rootDirectory: dir, fileList: [] }))
+    const filesToOpen = Array.from(fileSet)
+
+    // Discard all directories except first one and add files.
+    if (openFilesInSameWindow) {
+      if (directoriesToOpen.length) {
+        directoriesToOpen[0].fileList.push(...filesToOpen)
+        directoriesToOpen.length = 1
+      } else {
+        directoriesToOpen.push({ rootDirectory: null, fileList: filesToOpen })
+      }
+      filesToOpen.length = 0
+    }
+
+    // Find the best window(s) to open the files in.
+    if (!openFilesInSameWindow && !openFilesInNewWindow) {
+      const isFirstWindow = _windowManager.getActiveEditorId() === null
+
+      // Prefer new directories
+      for (let i = 0; i < directoriesToOpen.length; ++i) {
+        const { fileList, rootDirectory } = directoriesToOpen[i]
+
+        for (let j = 0; j < filesToOpen.length; ++j) {
+          const pathname = filesToOpen[j]
+          if (isChildOfDirectory(rootDirectory, pathname)) {
+            if (isFirstWindow) {
+              fileList.push(...filesToOpen)
+              filesToOpen.length = 0
+              i = 9999
+              break
+            }
+            fileList.push(pathname)
+            filesToOpen.splice(j, 1)
+            --j
+          }
+        }
+      }
+
+      // Find for the remaining files the best window to open the files in.
+      if (isFirstWindow && directoriesToOpen.length && filesToOpen.length) {
+        const { fileList } = directoriesToOpen[0]
+        fileList.push(...filesToOpen)
+        filesToOpen.length = 0
+      } else {
+        const windowList = _windowManager.findBestWindowToOpenIn(filesToOpen)
+        for (const item of windowList) {
+          const { windowId, fileList } = item
+
+          // File list is empty when all files are already opened.
+          if (fileList.length === 0) {
+            continue
+          }
+
+          if (windowId !== null) {
+            const window = _windowManager.get(windowId)
+            if (window) {
+              window.openTabs(fileList, 0)
+              window.bringToFront()
+              continue
+            }
+            // else: fallthrough
+          }
+          this._createEditorWindow(null, fileList)
+        }
+      }
+
+      // Directores are always opened in a new window if not already opened.
+      for (const item of directoriesToOpen) {
```

Comment:
`directoriesToOpen `'s length will always be one?

Reply:
Nope, please see above.